### PR TITLE
test: enable and fix Hardhat eth_call tests

### PR DESCRIPTION
**Pull request by Wodann**

### CODE REVIEW

File: crates/edr_provider/src/requests/eth/call.rs [hunk 1]
```diff
@@ -70,7 +61,10 @@ fn resolve_call_request(
         })
     } else {
         // Question: Is there a reason we don't support EIP-2930 calls?
```
Comment (Wodann) [inline crates/edr_provider/src/requests/eth/call.rs:63]: @fvictorio & @alcuadrado?

File: crates/edr_provider/src/pending.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+};
+
+/// A blockchain with a pending block.
+///
+/// # Panics
+///
+/// Panics if a state override is provided to `state_at_block_number` for the
+/// pending block; or if the `BlockchainMut` methods are called.
+#[derive(Debug)]
+pub(crate) struct BlockchainWithPending<'blockchain> {
```
Comment (agostbiro) [inline crates/edr_provider/src/pending.rs:18]: I'm a little bit concerned about the various asserts and panics and I'm wondering if we really need this type?

As far as I can tell, we only need `BlockchainWithPending` for the `dry_run` in `run_call` and `dry_run` only uses `spec_at_block_number` and `BlockHashRef`. So an alternative option could be to introduce an enum that is either a `SyncBlockchain` or a `SyncBlockchain` + a pending block, implement `spec_at_block_number` and `BlockHashRef` for this enum and have `dry_run` take this enum as argument instead of `SyncBlock`.

File: packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts [hunk 3]
```diff
@@ -603,7 +603,8 @@ describe("Eth module", function () {
             );
           });
 
-          it("should throw an error because an invalid storage key is used", async function () {
+          // TODO: https://github.com/NomicFoundation/edr/issues/104
+          it.skip("should throw an error because an invalid storage key is used", async function () {
```
Comment (agostbiro) [inline packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts:607]: This one returns a value instead of throwing. I think we should fix that before merging.

File: packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts [hunk 4]
```diff
@@ -581,7 +580,8 @@ describe("Eth module", function () {
             );
           });
 
-          it("should throw an error because the key address used is invalid", async function () {
+          // TODO: https://github.com/NomicFoundation/edr/issues/104
+          it.skip("should throw an error because the key address used is invalid", async function () {
             await assertInvalidArgumentsError(
```
Comment (agostbiro) [inline packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts:585]: @fvictorio I'm wondering if it'd be ok to add a flag to `assertInvalidArgumentsError` to only expect an error in EDR mode and not match the message so that we can have these enabled?

**Review by agostbiro — CHANGES_REQUESTED** [review]

File: crates/edr_provider/src/pending.rs (hunk 2 above)
Comment (Wodann) [reply crates/edr_provider/src/pending.rs:18]: I agree. Following your feedback, I rewrote the code such that the `runtime.rs` run-methods all only need the `BlockHashRef` trait, removing the need for the other implementations on `BlockchainWithPending`.

This is where I ran into trouble. We need to _upcast_ a dynamic trait object, which is currently not natively supported in Rust. Rust v1.76.0 will introduce this and is scheduled for release on February 8th, 2024.

To still fulfil your request, I tried all recommended workarounds for this issue that I could find, but none of them worked. After spending several hours on this, I feel that it's not worth spending more time on until native dynamic trait object upcasting is supported in Rust. I created [this issue](https://github.com/NomicFoundation/edr/issues/244) for it.

I believe that the documentation on `BlockchainWithPending` and is locality (`pub(crate)` should be sufficient to avoid usage with the unsupported trait implementations.

File: crates/edr_evm/src/runtime.rs [hunk 5]
```diff
@@ -50,6 +50,8 @@ where
 
     let state_overrider = StateRefOverrider::new(state_overrides, &state);
 
+    cfg.disable_balance_check = true;
+    cfg.disable_block_gas_limit = true;
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm/src/runtime.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4666/commits/5bd17bbf0980b725873890b7230aa28b9e317a32

File: crates/edr_provider/src/data.rs [hunk 6]
```diff
@@ -624,21 +665,22 @@ impl ProviderData {
         Ok(())
     }
 
+    // TransactionCreationError::NonceTooLow {
+    //     transaction_nonce, ..
+    // } => ProviderError::AutoMineNonceTooLow {
+    //     expected: next_nonce,
+    //     actual: transaction_nonce,
+    // }
+
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_provider/src/data.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4666/commits/5bd17bbf0980b725873890b7230aa28b9e317a32

File: packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts (hunk 3 above)
Comment (Wodann) [reply packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts]: Fixed by https://github.com/NomicFoundation/hardhat/pull/4666/commits/2aa30168e8246a9e76cb4a64f872d13ee8ece3dc

File: packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts (hunk 4 above)
Comment (Wodann) [reply packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/call.ts:592]: Fixed by https://github.com/NomicFoundation/hardhat/pull/4666/commits/a24682b8fa5e9a45c0b2cfd65dd9431bd4b9b43f and https://github.com/NomicFoundation/hardhat/pull/4666/commits/728ee6af1efbe2766c4c23d15e8818dfbe396032

File: crates/edr_provider/src/requests/eth/call.rs (hunk 1 above)
Comment (fvictorio) [reply crates/edr_provider/src/requests/eth/call.rs:63]: I guess we should, even if they are uncommon? My understanding is that an EIP-2930 transaction request should be used if the call object has both `gasPrice` and `accessList` fields.

But as always, if Hardhat is not currently doing this, then I think it's fine to open an issue for it and keep it as it is. An EIP-2930 seems like a very edge case to me.

**Review by agostbiro — CHANGES_REQUESTED** [review]
It looks like this branch has a regression for send transaction tests. Could you fix that @Wodann  before merging please? Otherwise LGTM.

File: crates/edr_provider/src/requests/eth/call.rs (hunk 1 above)
Comment (Wodann) [reply crates/edr_provider/src/requests/eth/call.rs:63]: https://github.com/NomicFoundation/edr/issues/255

**Review by agostbiro — APPROVED** [review]
